scripts/easy_train.py

```diff
@@ -0,0 +1,1844 @@
+                    lr=args.lr,
+                    gamma=args.gamma,
+                    network_save_period=args.network_save_period,
+                    save_last_network=args.save_last_network,
+                    seed=args.seed + run_id,
+                    start_from_model=start_model,
+                    root_dir=os.path.join(experiment_directory, 'training', f'run_{run_id}'),
+                    epoch_size=args.epoch_size,
+                    validation_size=args.validation_size,
+                    resume_training=args.resume_training,
+                    additional_args=[arg for arg in args.additional_training_args or []]
+                ))
+        LOGGER.info(f'Doing network training on gpus {gpu_ids}. {len(training_runs)} runs in total.')
+    else:
+        LOGGER.info('Not training networks.')
+
+    network_testing = NetworkTesting(
+        nnue_pytorch_directory=nnue_pytorch_directory,
+        root_dir=os.path.join(experiment_directory, 'training'),
+        ordo_exe=None if args.do_approximate_ordo else make_ordo_executable_path(ordo_directory),
+        c_chess_cli_exe=make_c_ches_cli_executable_path(c_chess_cli_directory),
+        stockfish_base_exe=make_stockfish_executable_path(stockfish_base_directory),
+        stockfish_test_exe=make_stockfish_executable_path(stockfish_test_directory),
+        features=args.features,
+        num_parallel_games=args.network_testing_threads,
+        explore_factor=args.network_testing_explore_factor,
+        book_file_path=args.network_testing_book,
+        time_per_game=args.network_testing_time_per_game,
+        time_increment_per_move=args.network_testing_time_increment_per_move,
+        nodes_per_move=args.network_testing_nodes_per_move,
+        hash=args.network_testing_hash_mb,
+        games_per_round=args.network_testing_games_per_round,
+        active=do_network_testing,
+        additional_args=[arg for arg in args.additional_testing_args or []]
+    )
+
+    #TODO it would be nice to be able to set a time limit at the cmdline and stop after e.g. 20h
+
+    for tr in training_runs:
+        tr.start()
+
+    network_testing.start()
+
+    if args.tui:
+        for h in LOGGER.handlers:
+            if isinstance(h, logging.StreamHandler):
+                LOGGER.removeHandler(h)
+        last_scene = None
+        while True:
+            try:
+                Screen.wrapper(app, catch_interrupt=True, arguments=[last_scene, training_runs, network_testing])
+                break
+            except ResizeScreenError as e:
+                last_scene = e.scene
+    else:
+        while True:
+            try:
+                v = input()
+                if v == 'quit':
+                    break
+                else:
+                    print('Type `quit` to stop.')
+            except EOFError:
+                # For non-interactive environments
+                time.sleep(1)
+
+    for tr in training_runs:
+        tr.stop()
+    network_testing.stop()
```

Comment:
stopping it at the same time results in latest networks being undertested. Usually I run testing for an hour or so after the training finishes. Not sure how to handle that best.

Reply:
Not sure either. I usually stop testing at the same time, meaning the last trained network is ignored from a testing point of view. It would be nice to have good termination criterion of easy_train. It could be max-epoch, which is rather simple, but maybe we need something more elaborate?